### DISCUSSION TITLE
Facing error while deploying OpenCRVS version 1.4

### DISCUSSION THREAD

**Discussion by sharma-rajat-93** [opencrvs/opencrvs-core#6605] [Technical Q&A] [answered]
### OpenCRVS Core version

1.4.0

### OpenCRVS Countryconfig version

1.4.0

### Your operating system and version

Ubuntu 22.04

### Your node version

v16.13.0

### Which of these describes your project best?

- [X] Testing OpenCRVS for business development, educational purposes or for personal interest.
- [X] A research project to evaluate, configure and test OpenCRVS as a possible solution for a government, an active brief or proposal.
- [ ] Actively installing and configuring OpenCRVS Core for production use in a country.
- [ ] Customising OpenCRVS Core code in production in a country

**Comment by sharma-rajat-93** [comment, marked as answer]
Hi Team, I am trying to deploy OpenCRVS v1.4 to a qa environment but when running the GitHub action script, I am facing this error:

![Screenshot 2024-03-13 at 1 00 41 PM](https://github.com/opencrvs/opencrvs-core/assets/136357138/1cce0b2c-ba61-42e4-9fae-d186224d78a5)

I have forked the country config to my own repo which is here:

https://github.com/radiantsolutionsorg/opencrvs-bahamas

Any help will be greatly appreciated.

**Reply by rikukissa** [reply]
Hey @sharma-rajat-93 ! I'm not sure if it's related to the issue but I see the provisioning pipeline has not been run for QA causing some other errors like 

```
Error response from daemon: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.
```
Could you try running that first and then rerunning the deploy pipeline?

**Reply by sharma-rajat-93** [reply]
Hey @rikukissa , thank you for that. I had missed this step from the documentation. After I ran the provisioning successfully (after a few tries :D), I am now able to deploy the QA environment successfully. Thank you again for your help.

**Reply by rikukissa** [reply]
Super, great to hear that! Yea, provisioning and deployments usually take a few tries to get right as everyone's environment is slightly different 😄 Let me know if you encounter any other issues!

**Reply by sharma-rajat-93** [reply]
My plan is to look into MOSIP integration with OpenCRVS as the next step. If you have any insights for me in that direction, that'd be helpful :)

**Reply by naftis** [reply]
Our documentation for MOSIP integration can be found at [documentation.opencrvs.org](https://documentation.opencrvs.org/technology/interoperability/national-id-client#installing-a-mosip-enabled-national-id-configuration) and we've recently published a custom country config repository for MOSIP integration, [opencrvs-mosip](https://github.com/opencrvs/opencrvs-mosip).

We're currently working with MOSIP to finish the country config repo, but it even in it's current state it's helpful for:
- setting up the docker-compose files: https://github.com/opencrvs/opencrvs-mosip/blob/2654bf76b5651202141238ed6f4bf0dcb00194b2/infrastructure/docker-compose.development-deploy.yml#L111
- setting up the country config endpoint which puts the mosipAid to the bundle: https://github.com/opencrvs/opencrvs-mosip/blob/2654bf76b5651202141238ed6f4bf0dcb00194b2/src/index.ts#L525
- customizing the forms with the <kbd>Authorize</kbd> -button: https://github.com/opencrvs/opencrvs-mosip/blob/2654bf76b5651202141238ed6f4bf0dcb00194b2/src/form/birth/index.ts#L267
- flowing MOSIP AID to your certificates: https://github.com/opencrvs/opencrvs-mosip/commit/66d441e8fee94a9256f64fbd64f211c244942af3 and https://github.com/opencrvs/opencrvs-mosip/blob/2654bf76b5651202141238ed6f4bf0dcb00194b2/src/data-seeding/certificates/source/India-birth-certificate-v2.svg?short_path=fb12ce9#L31

**Reply by sharma-rajat-93** [reply]
Thank you @naftis for these helpful links. I will look into these in detail, but have a quick question before that. I know that to register openCRVS as a partner with MOSIP platform, MOSIP will ask me for a partner certificate. Could you point me to how can I generate that or do any of these links mention that step?

**Reply by naftis** [reply]
For our recent integrations MOSIP has generated the partner certificate for us

**Reply by sharma-rajat-93** [reply]
Hi @naftis , can you also please clarify on a few questions:

1. I checked the mosiptokenseeder block and it mentions TOKENSEEDER_CRYPTO_ENCRYPT__ENCRYPT_CERT_PATH and TOKENSEEDER_CRYPTO_SIGNATURE__SIGN_P12_FILE_PATH. These certificates will be provided by MOSIP? If yes, are these the partner certificate related files?
2. Do I also need to deploy the  MOSIP Mediator (https://github.com/opencrvs/mosip-mediator) as a separate service or will it be deployed as part of the deployment once I add the token seeder block to my deployment file?

**Reply by naftis** [reply]
1. Yes, right, they are provided by MOSIP. They are related to the auth partner certificate and key (keystore.p12) and auth partner certificate (ida.partner.crt / .key). These are more MOSIP's expertise, and for further clarification, you should enquire with them.
2. You need to deploy it as a separate service, as being shown under the token seeder block: https://github.com/opencrvs/opencrvs-mosip/blob/2654bf76b5651202141238ed6f4bf0dcb00194b2/infrastructure/docker-compose.development-deploy.yml#L134

**Reply by sharma-rajat-93** [reply]
Ok thank you @naftis , I see that the docker compose yaml file also references these variables:

      - NATIONAL_ID_OIDP_BASE_URL=${NATIONAL_ID_OIDP_BASE_URL}
      - NATIONAL_ID_OIDP_REST_URL=${NATIONAL_ID_OIDP_REST_URL}
      - NATIONAL_ID_OIDP_CLIENT_PRIVATE_KEY=${NATIONAL_ID_OIDP_CLIENT_PRIVATE_KEY}
      - NATIONAL_ID_OIDP_JWT_AUD_CLAIM=${NATIONAL_ID_OIDP_JWT_AUD_CLAIM}

I didn't find anything in the documentation that speaks to these values. Can you please elaborate on what these values are and where am I supposed to get these?


FYI I got this error while deploying the development environment with MOSIP:

![Screenshot 2024-03-18 at 7 13 23 PM](https://github.com/opencrvs/opencrvs-core/assets/136357138/2504deff-2aab-488a-a33f-277aa96b1031)

**Reply by naftis** [reply]
These are the values that connect to a OIDP integration, in MOSIP's case it's E-Signet. The values could be for example:
```
NATIONAL_ID_OIDP_BASE_URL=https://esignet.test.mosip.net/
NATIONAL_ID_OIDP_REST_URL=https://esignet.test.mosip.net/v1/esignet/
NATIONAL_ID_OIDP_JWT_AUD_CLAIM=https://esignet.test.mosip.net/v1/esignet/oauth/v2/token
NATIONAL_ID_OIDP_CLIENT_ID=<<mosip provided client_id>>
NATIONAL_ID_OIDP_ESSENTIAL_CLAIMS=name,birthdate,address
NATIONAL_ID_OIDP_CLIENT_PRIVATE_KEY=<<base64 encoded private key, provided by MOSIP>>
```

The mediator client id, client secret and SHA secret are the national ID system integration values created via the national system admin UI in OpenCRVS. MOSIP provides the rest of the values; auth client id and secret being auth partner values, auth URL being similar to `https://iam.test.mosip.net/auth/realms/mosip/protocol/openid-connect/token`, auth user & pass being the keystore partner user name and password. The death and birth proxy URL and generate AID url are URL's that are running in MOSIP-side mediator.

We're looking into improving the documentation as we speak, and your feedback is valuable :)

**Reply by sharma-rajat-93** [reply]
Okay I’ll get the MOSIP side values but to get the mediator client id, secret etc. from the openCRVS UI, I’ll need to deploy it first correct? So before deploying successfully,  how can I get these values beforehand?

**Comment by sharma-rajat-93** [comment]
Okay I’ll get the MOSIP side values but to get the mediator client id,
secret etc. from the openCRVS UI, I’ll need to deploy it first correct? So
while deploying successfully,  how can I get these values beforehand?

**Reply by naftis** [reply]
You can first deploy OpenCRVS without the mediator, either adding placeholder values for environment variables or not deploying it. The mediator is a separate piece of software that just utilizes the OpenCRVS integrations.

**Reply by sharma-rajat-93** [reply]
There is a step in the MOSIP side mediator when deploying, which requires some certificate from OpenCRVS. This is the step listed there:
```Get certificate from OpenCRVS.```

Can you please help me understand what is this certificate and where can I get it?

**Reply by naftis** [reply]
From the MOSIP's install.sh, [would this refer to OpenCRVS public key](https://github.com/mosip/mosip-opencrvs/blob/develop/deployment/install.sh#L47) which can be found from your installation's auth service: `https://auth.{{hostname}}/.well-known`

**Reply by sharma-rajat-93** [reply]
ok thank you, I was able to download it.
